### PR TITLE
CDAP-13274 Emit FLL from Spanner source and sinks

### PR DESCRIPTION
**Pull request by rohitsinha54**
- Emit FLL from Spanner source and sink

Source:
image

Sink (Operations screenshot)
image

### CODE REVIEW

**Comment by rohitsinha54** [comment]
@albertshau The operation name in the second screenshot for source and sink are not consistent. The source and sink have the same full descriptive name in the plugin json. Although the source plugin which gets created when pipeline is created from Wrangler does not have full name and just have 'Spanner' as a name. Do you know where it might be coming from?

**Comment by albertshau** [comment]
The first part is the stage name in the pipeline I believe

File: src/main/java/co/cask/gcp/spanner/sink/SpannerSink.java [hunk 1]
```diff
@@ -72,6 +81,16 @@ public void prepareRun(BatchSinkContext context) throws Exception {
     SpannerOutputFormat.configure(configuration, config);
     context.addOutput(Output.of(config.referenceName,
                                 new SinkOutputFormatProvider(SpannerOutputFormat.class, configuration)));
+
+    List<Schema.Field> fields = config.getSchema().getFields();
+    if (fields != null && !fields.isEmpty()) {
+        // Record the field level WriteOperation
+        FieldOperation operation = new FieldWriteOperation("Write", "Wrote to SpannerSink table.",
```
Comment (albertshau) [inline src/main/java/co/cask/gcp/spanner/sink/SpannerSink.java]: Wrote to Spanner table.

File: src/main/java/co/cask/gcp/spanner/source/SpannerSource.java [hunk 2]
```diff
@@ -132,6 +137,16 @@ public void prepareRun(BatchSourceContext batchSourceContext) throws Exception {
     // set input format and pass configuration
     batchSourceContext.setInput(Input.of(config.referenceName,
                                          new SourceInputFormatProvider(SpannerInputFormat.class, configuration)));
+    if (config.getSchema() != null) {
+      if (config.getSchema().getFields() != null) {
+        FieldOperation operation =
+          new FieldReadOperation("Read", "Read from SpannerSource table",
```
Comment (albertshau) [inline src/main/java/co/cask/gcp/spanner/source/SpannerSource.java]: Read from Spanner table.

**Review by albertshau — APPROVED** [review]
minor comments, otherwise lgtm. fix and merge.

**Pull request merged**